lib/babe/build.go

```diff
@@ -237,6 +237,147 @@ func (b *BlockBuilder) buildBlockExtrinsics(slot Slot, rt runtime.Instance) []*t
+	Signature []byte // CollatorSignature
+	// The hash of the `pov-block`.
+	PovHash common.Hash
+}
+
+// UpwardMessage is a message from a parachain to its Relay Chain.
+type UpwardMessage []byte
+
+// OutboundHrmpMessage is an HRMP message seen from the perspective of a sender.
+type OutboundHrmpMessage struct {
+	Recipient uint32
+	Data      []byte
+}
+
+// All Vec<u8> in rust have become []byte here
+
+// ValidationCode is Parachain validation code.
+type ValidationCode []byte
+
+// HeadData is Parachain head data included in the chain.
+type HeadData []byte
+
+// CandidateCommitments are Commitments made in a `CandidateReceipt`. Many of these are outputs of validation.
+type CandidateCommitments struct {
+	// Messages destined to be interpreted by the Relay chain itself.
+	UpwardMessages []UpwardMessage
+	// Horizontal messages sent by the parachain.
+	HorizontalMessages []OutboundHrmpMessage
+	// New validation code.
+	NewValidationCode *ValidationCode
+	// The head-data produced as a result of execution.
+	HeadData HeadData
+	// The number of messages processed from the DMQ.
+	ProcessedDownwardMessages uint32
+	// The mark which specifies the block number up to which all inbound HRMP messages are processed.
+	HrmpWatermark uint32
+}
+
+// CommittedCandidateReceipt is a candidate-receipt with commitments directly included.
+type CommittedCandidateReceipt struct {
+	Descriptor  *CandidateDescriptor
+	Commitments *CandidateCommitments
+}
+
+// ValidityAttestation is an implicit or explicit attestation to the validity of a parachain
+// candidate.
+type ValidityAttestation int
+
+const (
+	// Implicit is for implicit attestation.
+	Implicit ValidityAttestation = iota
+	// Explicit is for explicit attestation.
+	Explicit
+)
+
+// UncheckedSignedAvailabilityBitfield is a set of unchecked signed availability bitfields.
+// Should be sorted by validator index, ascending.
+type UncheckedSignedAvailabilityBitfield struct {
+	// The payload is part of the signed data. The rest is the signing context,
+	// which is known both at signing and at validation.
+	Payload []byte
+	// The index of the validator signing this statement.
+	ValidatorIndex uint32
+	/// The signature by the validator of the signed payload.
+	Signature []byte
+}
+
+// BackedCandidate is a backed (or backable, depending on context) candidate.
+type BackedCandidate struct {
+	// The candidate referred to.
+	Candidate *CommittedCandidateReceipt
+	// The validity votes themselves, expressed as signatures.
+	ValidityVotes []*ValidityAttestation
+	// The indices of the validators within the group, expressed as a bitfield.
+	ValidatorIndices []byte
+}
+
+// MultiDisputeStatementSet is a set of dispute statements.
+type MultiDisputeStatementSet []DisputeStatementSet
+
+// DisputeStatement is a statement about a candidate, to be used within the dispute
+// resolution process. Statements are either in favour of the candidate's validity
+// or against it.
+type DisputeStatement int
```

Comment:
nit to take less space

```suggestion
type DisputeStatement uint8
```